src/lib.rs

```diff
@@ -21,24 +22,288 @@ pub use wgt::{
     VertexAttributeDescriptor, VertexFormat, BIND_BUFFER_ALIGNMENT, MAX_BIND_GROUPS,
 };
 
-//TODO: avoid heap allocating vectors during resource creation.
-#[derive(Default, Debug)]
-struct Temp {
-    //bind_group_descriptors: Vec<wgn::BindGroupDescriptor>,
-//vertex_buffers: Vec<wgn::VertexBufferDescriptor>,
+use backend::Context as C;
+
+trait ComputePassInner<Ctx: Context> {
+    fn set_pipeline(&mut self, pipeline: &Ctx::ComputePipelineId);
+    fn set_bind_group(
+        &mut self,
+        index: u32,
+        bind_group: &Ctx::BindGroupId,
+        offsets: &[DynamicOffset],
+    );
+    fn dispatch(&mut self, x: u32, y: u32, z: u32);
+    fn dispatch_indirect(
+        &mut self,
+        indirect_buffer: &Ctx::BufferId,
+        indirect_offset: BufferAddress,
+    );
+}
+
+trait RenderPassInner<Ctx: Context> {
+    fn set_pipeline(&mut self, pipeline: &Ctx::RenderPipelineId);
+    fn set_bind_group(
+        &mut self,
+        index: u32,
+        bind_group: &Ctx::BindGroupId,
+        offsets: &[DynamicOffset],
+    );
+    fn set_index_buffer(
+        &mut self,
+        buffer: &Ctx::BufferId,
+        offset: BufferAddress,
+        size: BufferAddress,
+    );
+    fn set_vertex_buffer(
+        &mut self,
+        slot: u32,
+        buffer: &Ctx::BufferId,
+        offset: BufferAddress,
+        size: BufferAddress,
+    );
+    fn set_blend_color(&mut self, color: wgt::Color);
+    fn set_scissor_rect(&mut self, x: u32, y: u32, width: u32, height: u32);
+    fn set_viewport(
+        &mut self,
+        x: f32,
+        y: f32,
+        width: f32,
+        height: f32,
+        min_depth: f32,
+        max_depth: f32,
+    );
+    fn set_stencil_reference(&mut self, reference: u32);
+    fn draw(&mut self, vertices: Range<u32>, instances: Range<u32>);
+    fn draw_indexed(&mut self, indices: Range<u32>, base_vertex: i32, instances: Range<u32>);
+    fn draw_indirect(&mut self, indirect_buffer: &Ctx::BufferId, indirect_offset: BufferAddress);
+    fn draw_indexed_indirect(
+        &mut self,
+        indirect_buffer: &Ctx::BufferId,
+        indirect_offset: BufferAddress,
+    );
+}
+
+trait Context: Sized {
```

Comment:
I think the traits should be in a different file so it's easier to read this one.

Reply:
It could be, yes. I find it convenient here, because it uses all the types in the proximity, and everything uses it. We can follow-up with moving it if we see fit, after the initial work is done.